test/out_of_order_engine_tests.cc

```diff
@@ -0,0 +1,562 @@
+
+	{
+		const auto iq = octx.assign_all();
+		CHECK(iq.assigned_instructions_are({}));
+	}
+
+	octx.complete(d1_k1);
+
+	{
+		const auto iq = octx.assign_all();
+		CHECK(iq.assigned_instructions_are({h0, h1, h2}));
+		CHECK(iq.assigned_in_order({h0, h1, h2}));
+	}
+}
+
+TEST_CASE("out_of_order_engine eagerly assigns copy-instructions to the lanes of their dependencies", "[out_of_order_engine]") {
+	out_of_order_test_context octx(4);
+	const auto d0_k0 = octx.device_kernel({}, device_id(0));
+	const auto d1_k0 = octx.device_kernel({}, device_id(1));
+	const auto d2_k0 = octx.device_kernel({}, device_id(2));
+	const auto d3_k0 = octx.device_kernel({}, device_id(3));
+	const auto copy_dep0 = octx.copy({d0_k0}, first_device_memory_id, first_device_memory_id + 1);
+	const auto copy_dep1 = octx.copy({d1_k0}, first_device_memory_id, first_device_memory_id + 1);
+	const auto copy_dep2 = octx.copy({d2_k0}, host_memory_id, first_device_memory_id + 2);
+	const auto copy_dep3 = octx.copy({d3_k0}, first_device_memory_id + 3, host_memory_id);
+	const auto copy_host = octx.copy({copy_dep3}, host_memory_id, host_memory_id); // can't be enqueued on a device in-order queue
+
+	CHECK_FALSE(octx.is_idle());
+
+	{
+		const auto iq = octx.assign_all();
+		CHECK(iq.assigned_instructions_are({d0_k0, d1_k0, d2_k0, d3_k0, copy_dep0, copy_dep1, copy_dep2, copy_dep3}));
+		CHECK(iq.queued_on_device(d0_k0, device_id(0)));
+		CHECK(iq.queued_on_device(d1_k0, device_id(1)));
+		CHECK(iq.queued_on_device(d2_k0, device_id(2)));
+		CHECK(iq.queued_on_device(d3_k0, device_id(3)));
+		CHECK(iq.assigned_in_order({d0_k0, copy_dep0}));
+		CHECK(iq.assigned_in_order({d1_k0, copy_dep1}));
+		CHECK(iq.assigned_in_order({d2_k0, copy_dep2}));
+		CHECK(iq.assigned_in_order({d3_k0, copy_dep3}));
+	}
+
+	octx.complete(d0_k0);
+	octx.complete(d1_k0);
+	octx.complete(d2_k0);
+	octx.complete(d3_k0);
+	CHECK_FALSE(octx.is_idle());
+
+	const auto copy_indep0 = octx.copy({d0_k0 /* already complete */}, first_device_memory_id + 1, first_device_memory_id);
+	const auto copy_indep1 = octx.copy({d1_k0 /* already complete */}, first_device_memory_id, first_device_memory_id + 1);
+	const auto copy_indep2 = octx.copy({d2_k0 /* already complete */}, first_device_memory_id + 1, first_device_memory_id + 2);
+	const auto copy_indep3 = octx.copy({d3_k0 /* already complete */}, first_device_memory_id, first_device_memory_id + 3);
+
+	{
+		const auto iq = octx.assign_all();
+		CHECK(iq.assigned_instructions_are({copy_indep0, copy_indep1, copy_indep2, copy_indep3}));
+		CHECK((iq.queued_on_device(copy_indep0, device_id(1)) || iq.queued_on_device(copy_indep0, device_id(0))));
+		CHECK((iq.queued_on_device(copy_indep1, device_id(0)) || iq.queued_on_device(copy_indep1, device_id(1))));
+		CHECK((iq.queued_on_device(copy_indep2, device_id(1)) || iq.queued_on_device(copy_indep2, device_id(2))));
+		CHECK((iq.queued_on_device(copy_indep3, device_id(0)) || iq.queued_on_device(copy_indep3, device_id(3))));
+	}
```

Comment:
What is this testing? It's not eager submission since the dependencies are already completed?

Reply:
Pretty much that: Copies are immediately assigned if their dependencies have completed before.